datagateway_api/common/config.py

```diff
@@ -109,25 +88,85 @@ def set_backend_type(self, backend_type):
         type must be fetched. This must be done using this module (rather than directly
         importing and checking the Flask app's config) to avoid circular import issues.
         """
-        self._config["backend"] = backend_type
+        self.backend = backend_type
 
-    def get_icat_properties(self):
+    class Config:
         """
-        ICAT properties can be retrieved using Python ICAT's client object, however this
-        requires the client object to be authenticated which may not always be the case
-        when requesting these properties, hence a HTTP request is sent as an alternative
+        The behaviour of the BaseModel class can be controlled via this class.
         """
-        properties_url = (
-            f"{config.get_config_value(APIConfigOptions.ICAT_URL)}/icat/properties"
-        )
-        r = requests.request(
-            "GET",
-            properties_url,
-            verify=config.get_config_value(APIConfigOptions.ICAT_CHECK_CERT),
-        )
-        icat_properties = r.json()
 
-        return icat_properties
+        # Enables assignment validation on the BaseModel fields. Useful for when the
+        # backend type is changed using the set_backend_type function.
+        validate_assignment = True
+
+
+class SearchAPI(BaseModel):
+    """
+    Configuration model class that implements pydantic's BaseModel class to allow for
+    validation of the SearchAPI config data using Python type annotations.
+    """
+
+    client_pool_init_size: StrictInt
+    client_pool_max_size: StrictInt
+    extension: StrictStr
+    icat_check_cert: StrictBool
+    icat_url: StrictStr
+
+
+class TestUserCredentials(BaseModel):
+    username: StrictStr
+    password: StrictStr
+
+
+class APIConfig(BaseModel):
+    """
+    Configuration model class that implements pydantic's BaseModel class to allow for
+    validation of the API config data using Python type annotations. It ensures that
+    all required config options exist before getting too far into the setup of the API.
+
+    If a mandatory config option is missing or misspelled, or has a wrong value type,
+    Pydantic raises a validation error with a breakdown of what was wrong and the
+    application is exited.
+
+    Config options used for testing are not checked here as they should only be used
+    during tests, not in the typical running of the API.
+
+    Some options used when running the API (host, debug_mode etc.) aren't mandatory
```

Comment:
Yes, as when the API is deployed for production use, they're set elsewhere (e.g. Apache is used to set the host and port). The following should be set to optional:
- `flask_reloader`
- `debug_mode`
- `host`
- `port`

I also wonder whether `test_user_credentials` and `test_mechanism` should be set as optional because these shouldn't be required for production use.

Reply:
Thanks! Yes, `test_user_credentials` and `test_mechanism` should also be optional